### ISSUE TITLE
Compiled stylesheet not using generated_images_dir

### ISSUE THREAD

**Issue by madhatter42** [Compass/compass#857]
I have an issue where I'm specifying a generated_images_dir path in my config file -- sprites are correctly generated to that folder, but the compiled stylesheet is not using that path for the sprites. I tried specifying a generated_images_path as well, but that makes no difference.

Here's my config file, I'm on version 0.12.1:

```
# Require any additional compass plugins here.

# Set this to the root of your project when deployed:
http_path = "/"
css_dir = "css"
sass_dir = "scss"
images_dir = "images"
javascripts_dir = "js"
generated_images_dir = 'images/compass-sprites'

# You can select your preferred output style here (can be overridden via the command line):
# output_style = :expanded or :nested or :compact or :compressed
output_style = (environment == :production) ? :compressed : :expanded

# To enable relative paths to assets via compass helper functions. Uncomment:
# relative_assets = true

# To disable debugging comments that display the original location of your selectors. Uncomment:
line_comments = (environment == :production) ? false : true

# If you prefer the indented syntax, you might want to regenerate this
# project again passing --syntax sass, or you can uncomment this:
# preferred_syntax = :sass
# and then run:
# sass-convert -R --from scss --to sass sass scss && rm -rf sass && mv scss sass
```

**Comment by mark-rushakoff** [comment]
If you're using compass-rails, it sounds like you might need to add `images/compass-sprites` to your asset path.  See [my blog post](http://pivotallabs.com/users/mrushakoff/blog/articles/2097-getting-compass-to-put-generated-sprites-somewhere-reasonable) or [this open pull request to compass-rails](https://github.com/Compass/compass-rails/pull/41).

**Comment by zgstyle** [comment]
+1 The same issue here, but I don't use compass-rails. It is standard Nanoc project.

**Comment by krisbulman** [comment]
Possibly related to https://github.com/chriseppstein/compass/issues/989 ?

**Comment by glortho** [comment]
+1, not using compass-rails.

**Comment by glortho** [comment]
Note: Fixed this in my environment by explicitly setting http_generated_images_path in the config file.

UPDATE: This seems to break the automatic removal of old sprites. This may be just by virtue of using the generated_images_dir directive...

**Comment by Frexuz** [comment]
+1 for non-rails project. I'm using sprite-path to calculate the width of the generated sprite file, which it cannot find.

**Comment by aprescott** [comment]
I'm having this same problem on a non-Rails project which uses as a stand-alone library. Despite setting all of the `{http_,}generated_images_{dir,path}` variables seemingly correctly, the generated CSS doesn't use the correct paths.

The source SCSS uses the `@import` and `my-foo-sprite(bar)` syntax [as described in the documentation](http://compass-style.org/help/tutorials/spriting/).

```
@import "my-icons/*.png";

.some-class {
  @include my-icons-sprite(new);
}
```

The configuration is set to

```
generated_images_dir = "foo/images/sprites"
generated_images_path = "foo/images/sprites"

http_generated_images_dir = "/images/sprites"
http_generated_images_path = "/images/sprites"
```

While images _do_ get created in `foo/images/sprites`, even though the `http_` config options are set, the generated CSS still uses `/images/my-icons/*.png` as though the generated dir/path settings aren't in effect.

**Comment by aprescott** [comment]
I think the issue in my case is that the `@import` uses an absolute path of `/my-icons/*.png` and it seems that `generated_images_url` doesn't prepend `generated_images_dir` if it's absolute:

``` ruby
def generated_image_url(path, cache_buster = Sass::Script::Bool.new(false))
  path = path.value # get to the string value of the literal.

  if path =~ %r{^#{Regexp.escape(Compass.configuration.http_generated_images_path)}/(.*)}
    # Treat root relative urls (without a protocol) like normal if they start with
    # the generated_images path.
    path = $1
  elsif absolute_path?(path)
    # Short curcuit if they have provided an absolute url.
    return Sass::Script::String.new("url(#{path})")
  end

  # ...
```

Because `http_generated_images_path` in my `config.rb` is actually `/images/sprites`, which differs from what `@import` is using, the `=~` regex check (given above) returns false and `absolute_path?` returns true.

Changing `@import` from `/my-icons/*.png` to `my-icons/*.png` and ensuring `images_dir` is set appropriately fixes my problem.

**Comment by dlindenkreuz** [comment]
On another note, in @aprescott's code sample above, the default value for `cache_buster` should be `Sass::Script::Bool.new(true)` so generated images such as sprites can also make use of modified URLs.

**Comment by alan-andrade** [comment]
@aprescott post fixed my problem :)  Thanks.

**Comment by bwiggs** [comment]
Did anyone ever figure out a solution for this? I'm experiencing the same issue mentioned in the original report.

**Comment by AlexSkrypnyk** [comment]
For future reference.
generated_images_dir is used when importing sprites and is also exposed for automatic inclusion, so need to specify parent "images" directory.

Files before generation:
root
|--images
**|--sprites-sources
__**|--*.png

In iconfig.rb:
images_dir = "images"
generated_images_dir = images_dir + "/sprites-generated"

In style.scss:
@import "sprites-sources/*.png";
@include all-sprites-sources-sprites;

Files after generation:
root
|--images
**|--sprites-sources
__**|--*.png
**|--sprites-generated
__**|--sprite-12321341234.png

If use full path from root when importing sprites (@import "images/sprites-sources/_.png"; instead of just @import "sprites-sources/_.png";) that would produce root->images>images>sprite-1234.png without respecting the directory set in generated_images_dir (/sprites-generated).

Files after incorrect generation:
root
|--images
**|--sprites-sources
__**|--*.png
**|--images
__**|--sprite-12321341234.png

**Comment by jonfaustman** [comment]
I just started having this problem with a stand-alone project after I updated RVM, Compass and Sass. My file dir. is a bit different:

template root
|--src/sass
|--src/img
|--static/2011/latest/css
|--static/2011/latest/img

and my config.rb:

css_dir = "static/2011/latest/css"
sass_dir = "src/sass"
fonts_dir = "static/2011/latest/fonts"
images_dir = "src/img"
generated_images_dir = "static/2011/latest/img"
javascripts_dir = "static/2011/latest/js"

So far not having any luck with generating dimensions at all.

**Comment by chriseppstein** [comment]
Pretty sure this is fixed in 1.0. Please retest.
